### PR TITLE
CA-228680: Improve error message when VM crashes before VM.start finishes

### PR DESCRIPTION
**Pull request by Frezzle**
Also fixed indentation.

Signed-off-by: Frederico Mazzone <frederico.mazzone@citrix.com>

### CODE REVIEW

**Comment by lindig** [comment]
This looks good. Maybe this is an opportunity to introduce better names for power states `a` and `b`? I'd call them `actual` and `expected` or similar.

**Comment by Frezzle** [comment]
@lindig You're right. I'll add that in.

**Comment by Frezzle** [comment]
@lindig Updated; I went with `found` and `expected` to match another use of this exception on line 2182.

**Comment by Frezzle** [comment]
Squashed into one commit.

**Comment by robhoes** [comment]
I'm not sure that this fix is right. This catches a check that that is done before the VM.start, but we wanted to fix the assertion afterwards:

    Xapi_vm_lifecycle.assert_power_state_is

File: ocaml/xapi/xapi_xenops.ml [hunk 1]
```diff
@@ -2896,6 +2894,7 @@ let vif_set_ipv4_configuration ~__context ~self =
        let module Client = (val make_client queue_name : XENOPS) in
        Client.VIF.set_ipv4_configuration dbg vif.Vif.id vif.Vif.ipv4_configuration |> sync_with_task __context queue_name;
        Events_from_xenopsd.wait queue_name dbg (fst vif.Vif.id) ();
+       Events_from_xenopsd.wait queue_name dbg (fst vif.Vif.id) ();
```
Comment (Frezzle) [inline ocaml/xapi/xapi_xenops.ml]: ..removed in new commit.

**Comment by Frezzle** [comment]
Updated. There were various places where `assert_power_state_is` and `assert_power_state_in` were used, but they were used in different ways; sometimes as a precheck to an operation, other times as a postcheck, and in 2 instances as part of another assertion (where those were also used as prechecks).

I therefore split the assert into 2 separate assert functions, where one is supposed to be used as a precheck (which throws `VM_BAD_POWER_STATE` to signify that the user has attempted something silly) and another to be used as a postcheck (which throws `INTERNAL_ERROR` to signify that something has gone wrong during an operation and lead to the VM ending up in an unexpected power state).

**Comment by Frezzle** [comment]
When ready to merge, please let me know and I can squash these commits.

File: ocaml/xapi/xapi_vm_lifecycle.ml [hunk 2]
```diff
@@ -602,17 +602,32 @@ let cancel_tasks ~__context ~self ~all_tasks_in_db ~task_ids =
...
+let assert_initial_power_state_in ~__context ~self ~allowed =
   let actual = Db.VM.get_power_state ~__context ~self in
   if not (List.mem actual allowed)
   then raise (Api_errors.Server_error(Api_errors.vm_bad_power_state, [
       Ref.string_of self;
       List.map Record_util.power_to_string allowed |> String.concat ";";
       Record_util.power_to_string actual ]))
 
-let assert_power_state_is ~expected = assert_power_state_in ~allowed:[expected]
+(** Assert if VM is in a certain state before starting an operation *)
```
Comment (lindig) [inline ocaml/xapi/xapi_vm_lifecycle.ml]: Not a native speaker, but should this not read "assert *that*"? This would also apply above.

Comment (Frezzle) [inline ocaml/xapi/xapi_vm_lifecycle.ml]: Yes, you're right; I've misused `assert` as a synonym for `check`. I'll change this.

**Comment by lindig** [comment]
I would like to get this merged because we see confusing error messages where a VM fails to start but the error message makes it sound like a user error.

File: ocaml/xapi/xapi_xenops.ml [hunk 3]
```diff
@@ -2516,8 +2517,7 @@ let shutdown ~__context ~self timeout =
            (fun () ->
               Events_from_xenopsd.wait queue_name dbg id ())
        in
-       Xapi_vm_lifecycle.assert_power_state_is ~__context ~self ~expected:`Halted;
-       (* force_state_reset called from the xenopsd event loop above *)
```
Comment (Frezzle) [inline ocaml/xapi/xapi_xenops.ml]: dont merge yet; somehow i removed this comment..

**Comment by Frezzle** [comment]
@robhoes Squashed and ready to merge.

**Pull request merged**